components/places/src/bookmark_sync/tests/payload_evolution.rs

```diff
@@ -0,0 +1,358 @@
+                guid('c'),
+                "Animals",
+                vec!["A", "B"],
+                vec![("device_type", "phone"), ("color", "blue")],
+            ),
+            remote_separator(guid('d'), 2, vec![("device_type", "desktop")]),
+            // We never change Query and Livemarks records, so we can't test roundtripping changes
+            // to those
+        ],
+        local_updates: vec![
+            LocalUpdate::BookmarkUpdate(title_change(guid('a'), "Doggies")),
+            LocalUpdate::BookmarkUpdate(title_change(guid('b'), "Kitties")),
+            LocalUpdate::BookmarkUpdate(title_change(guid('c'), "Cute Animals")),
+            LocalUpdate::BookmarkUpdate(pos_change(guid('d'), 3)),
+        ],
+        incoming_remote_records: vec![],
+        outgoing_unknown_fields: vec![
+            (guid('a'), vec![("device_type", "phone")]),
+            (guid('b'), vec![]),
+            (guid('c'), vec![("device_type", "phone"), ("color", "blue")]),
+            (guid('d'), vec![("device_type", "desktop")]),
+        ],
+    }
+    .run()
+}
+
+#[test]
+fn test_new_unknown_fields() {
+    // If we have new incoming remote records with new unknown fields, those should override the
+    // ones from the mirror table
+    RoundtripTest {
+        initial_remote_records: vec![
+            remote_bookmark(guid('a'), "Dogs", vec![("device_type", "phone")]),
+            remote_bookmark(guid('b'), "Cats", vec![("device_type", "desktop")]),
+        ],
+        local_updates: vec![
+            LocalUpdate::BookmarkUpdate(title_change(guid('a'), "Doggies")),
+            LocalUpdate::BookmarkUpdate(title_change(guid('b'), "Kitties")),
+        ],
+        incoming_remote_records: vec![remote_bookmark(
+            guid('a'),
+            "Dogs",
+            vec![("device_type", "mini-phone")],
+        )],
+        outgoing_unknown_fields: vec![
+            (guid('a'), vec![("device_type", "mini-phone")]),
+            (guid('b'), vec![("device_type", "desktop")]),
+        ],
+    }
+    .run()
+}
+
+#[test]
+fn test_reupload_after_reset() {
+    // After a reset, we re-upload all local records.  Those records should still have the unkown
+    // fields
+    RoundtripTest {
+        initial_remote_records: vec![
+            remote_bookmark(guid('a'), "Dogs", vec![("device_type", "phone")]),
+            remote_folder(
+                guid('b'),
+                "Animals",
+                vec!["A", "B"],
+                vec![("device_type", "phone"), ("color", "blue")],
+            ),
+            remote_separator(guid('c'), 2, vec![("device_type", "desktop")]),
+            // We re-upload query and livemark records after a reset, so test them here
+            // ... except for some reason apply_incoming() flags these as deleted and wants to sync
+            // outgoing tombstones for them
+            // remote_query(guid('d'), "dog search", "place:folder=123&excludeItems=1", vec![("device_type", "desktop")]),
```

Comment:
Here's the test fix for queries (plus plumbing `store_incoming_query` / `BookmarkItemRecord::unknown_fields`):

```suggestion
            remote_item(
                json!({
                    "id": "unfiled",
                    "title": "Unfiled Bookmarks",
                    "type": "folder",
                    "parentid": "places",
                    "parentName": "places",
                    "dateAdded": 0,
                    "children": vec![guid('d')],
                }),
                vec![],
            ),
            remote_query(guid('d'), "dog search", "place:folder=123&excludeItems=1", vec![("device_type", "desktop")]),
```

I'll post the explanation in a bit!

Reply:
Queries are still supported in Desktop Places. Fenix and iOS don't show them in the UI, but Rust Places still stores and round-trips them. There are three special cases in Sync's handling of queries:

1. Rewriting tag folder queries to reference tag names instead of folders.
2. Rewriting non-tag `folder` queries to add `excludeItems`.
3. Removing suspected "orphaned left pane queries".

I think your test unintentionally triggered (3), which is why it was uploading a tombstone.

What's a left pane query? If you open Bookmarks > Manage Bookmarks in Firefox Desktop, you'll see a two-pane window, with "All Bookmarks", "Bookmarks Toolbar", "Bookmarks Menu", "Other Bookmarks", and "Mobile Bookmarks" in the left pane. These left-pane queries show the contents of those roots in Places. Since 2017-ish, these queries are implemented entirely in the UI.

Before 2017, these left-pane queries were stored in Places as real query items, as children and grandchildren of the "left pane root". This root was hidden from the UI, and not synced. Before our Desktop bookmarks tracker and mirror rewrites, though, it was possible in some cases for Sync to observe changes to those left pane queries–like localized titles, or query URLs, changing between releases—and upload records for them. Crucially, it never uploaded the left pane root, so receiving devices would see those left pane queries, decide they were orphaned, and move them into `unfiled`.

This caused all sorts of issues in the UI. In the worst case, expanding one of those orphaned left pane queries at the wrong time would cause an infinite loop. We didn't have a good way of identifying these queries—their URLs and titles could change between releases, and the left pane root had a random GUID per device—so we used a [heuristic](https://github.com/mozilla/dogear/blob/f83e6db9640332bb4d8b9689740a7db68ba087f3/src/tree.rs#L1583-L1584). **If we saw any orphaned query—where it had a `parentid` that either didn't exist, or didn't mention the query in its `children` array—we'd delete it.**

TL;DR: Your test was accidentally creating one of these orphaned queries: `remote_query()` returns a record with `"parentid: "unfiled"`, but `initial_remote_records` didn't have a record for `unfiled`, so Dogear assumed it was an orphaned left pane query, and deleted it.

Your test didn't trigger the other two special cases, but here's the explanation for completeness.

(1) and (2) have the same underlying cause: the `folder` param references the SQLite row ID, which isn't portable across devices. When a query references a nonexistent `folder` ID, it'll return _all_ bookmarks in the database, which breaks the UI in the same way as (3).

To fix (1), Sync rewrites tag queries (with `type=7`) from `folder=<tag folder ID>` to `tag=<tag name>`. Desktop Places added support for `tag=` params in Firefox 61, and Sync query records store the tag name in the `folderName` in their payloads, so it has enough information to fix these up.

Sync doesn't have enough information to fix up (2)—the receiving device doesn't know what folder ID the query wants—so it compromises by adding `excludeItems=1` to the query. The device that originally uploaded the query has to rewrite it from `folder=<folder ID> to `parent=<folder GUID>`, but, in the meantime, adding `excludeItems=1` will make sure the query doesn't return anything, to keep it from breaking the UI.

Reply:
Wow, that history is totally wild.  I copied your code and was able to get the remote_query test to fail properly and then was able to fix it.  I think payload evolution is working for all bookmark types other than livemarks, thanks for all the help!